text/0000-debug-map-key-value.md

```diff
@@ -0,0 +1,265 @@
+
+Why would you want to use `key` and `value` directly if they're less convenient than `entry`? The reason is when the driver of the `DebugMap` is a framework like `serde` rather than a data structure directly:
+
+```rust
+struct DebugMap<'a, 'b: 'a>(fmt::DebugMap<'a, 'b>);
+
+impl<'a, 'b: 'a> SerializeMap for DebugMap<'a, 'b> {
+    type Ok = ();
+    type Error = Error;
+
+    fn serialize_key<T: ?Sized>(&mut self, key: &T) -> Result<(), Self::Error>
+    where
+        T: Serialize,
+    {
+        self.0.key(&key.to_debug());
+        Ok(())
+    }
+
+    fn serialize_value<T: ?Sized>(&mut self, value: &T) -> Result<(), Self::Error>
+    where
+        T: Serialize,
+    {
+        self.0.value(&value.to_debug());
+        Ok(())
+    }
+
+    fn serialize_entry<K: ?Sized, V: ?Sized>(
+        &mut self,
+        key: &K,
+        value: &V,
+    ) -> Result<(), Self::Error>
+    where
+        K: Serialize,
+        V: Serialize,
+    {
+        self.0.entry(&key.to_debug(), &value.to_debug());
+        Ok(())
+    }
+
+    fn end(self) -> Result<Self::Ok, Self::Error> {
+        self.0.finish().map_err(Into::into)
+    }
+}
+```
+
+Consumers should prefer calling `entry` over `key` and `value`.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The `key` and `value` methods can be implemented on `DebugMap` by tracking the state of the current entry in a `bool`, and splitting the existing `entry` method into two:
+
+```rust
+pub struct DebugMap<'a, 'b: 'a> {
+    has_key: bool,
+    ..
+}
+
+pub fn debug_map_new<'a, 'b>(fmt: &'a mut fmt::Formatter<'b>) -> DebugMap<'a, 'b> {
+    DebugMap {
+        has_key: false,
+        ..
+    }
+}
+
+impl<'a, 'b: 'a> DebugMap<'a, 'b> {
+    pub fn entry(&mut self, key: &dyn fmt::Debug, value: &dyn fmt::Debug) -> &mut DebugMap<'a, 'b> {
+        self.key(key).value(value)
+    }
+
+    pub fn key(&mut self, key: &dyn fmt::Debug) -> &mut DebugMap<'a, 'b> {
+        self.result = self.result.and_then(|_| {
+            // Make sure there isn't a partial entry
+            if self.has_key {
+                return Err(fmt::Error);
+            }
+
+            // write the key
+
+            // Mark that we're in an entry
+            self.has_key = true;
+            Ok(())
+        });
+
+        self
+    }
+
+    pub fn value(&mut self, value: &dyn fmt::Debug) -> &mut DebugMap<'a, 'b> {
+        self.result = self.result.and_then(|_| {
+            // Make sure there is a partial entry to finish
+            if !self.has_key {
+                return Err(fmt::Error);
+            }
+
+            // write the value
+
+            // Mark that we're not in an entry
+            self.has_key = false;
+            Ok(())
+        });
+
+        self.has_fields = true;
+        self
+    }
+
+    pub fn finish(&mut self) -> fmt::Result {
+        // Make sure there isn't a partial entry
+        if self.has_key {
+            return Err(fmt::Error);
```

Comment:
Creating a `fmt::Error` like this isn't allowed. From https://doc.rust-lang.org/nightly/std/fmt/index.html#formatting-traits:
> a formatting implementation must and may only return an error if the passed-in Formatter returns an error.

As this indicates a programming error I think an explicit `panic` here would be correct.

Reply:
Sure! A panic seems reasonable.